rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs

```diff
@@ -0,0 +1,467 @@
+            minimum_throughput,
+            inner: body,
+            sleep_fut: None,
+        }
+    }
+}
+
+#[cfg(feature = "http-0-x")]
+impl<B> http_body::Body for MinimumThroughputBody<B>
+where
+    B: http_body::Body<Data = bytes::Bytes, Error = BoxError>,
+{
+    type Data = bytes::Bytes;
+    type Error = BoxError;
+
+    fn poll_data(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Self::Data, Self::Error>>> {
+        let now = self.time_source.now();
+        let mut a_log_was_pushed_this_poll = false;
+        // Attempt to read the data from the inner body, then update the
+        // throughput logs.
+        let mut this = self.as_mut().project();
+        // Push a start log if we haven't already done so.
+        if this.throughput_logs.is_empty() {
+            this.throughput_logs.push((now, 0));
+            a_log_was_pushed_this_poll = true;
+        }
+        let poll_res = match this.inner.poll_data(cx) {
+            Poll::Ready(Some(Ok(bytes))) => {
+                this.throughput_logs.push((now, bytes.len() as u64));
+                a_log_was_pushed_this_poll = true;
+                Poll::Ready(Some(Ok(bytes)))
+            }
+            Poll::Pending => Poll::Pending,
+            // If we've read all the data or an error occurred, then return that result.
+            res => return res,
+        };
+
+        // Check the sleep future to see if it needs refreshing.
+        let mut sleep_fut = this.sleep_fut.take().unwrap_or_else(|| {
+            this.async_sleep
+                .sleep(this.minimum_throughput.per_time_elapsed())
+        });
+        if let Poll::Ready(()) = pin!(&mut sleep_fut).poll(cx) {
+            // Whenever the sleep future expires, we replace it.
+            sleep_fut = this
+                .async_sleep
+                .sleep(this.minimum_throughput.per_time_elapsed());
+            // If we already pushed a log during this poll, don't set a
+            // "floating back". Otherwise, do set one.
+            //
+            // The floating back acts as a end-bound for the purposes of
+            // calculating throughput for a given time span. 0-byte logs
+            // don't affect the throughput calculation except when they are
+            // the first or last log. The starting log is pushed the first
+            // time this is polled. If no data was emitted during a poll,
+            // then we set the floating back.
+            if !a_log_was_pushed_this_poll {
+                this.throughput_logs.set_floating_back(now);
+            }
+            // We also schedule a wake up for current task to ensure that
+            // it gets polled at least one more time.
+            cx.waker().wake_by_ref();
+        };
+        this.sleep_fut.replace(sleep_fut);
+
+        // Calculate the current throughput and emit an error if it's too low.
+        let actual_throughput = this.throughput_logs.calculate_throughput()?;
```

Comment:
is it worth benchmarking this at some point? if we're getting a ton of data how much time do we spend calculating throughput? I wonder if we should _only_ calculate it when the sleep expires, e.g.

Reply:
Agreed, just not in this PR.